File: DSA/Yashvardhan-Guliyani/Day6_q1.java
class ListNode {
    int val;
    ListNode next;
    
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class LinkedListCycleDetection {
    public static String hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return "no cycle";
        }
        
        ListNode slow = head;
        ListNode fast = head.next;
        
        while (slow != fast) {
            if (fast == null || fast.next == null) {
                return "no cycle";
            }
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return "cycle";
    }
    
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode second = new ListNode(2);
        ListNode third = new ListNode(3);
        ListNode fourth = new ListNode(4);
        
        head.next = second;
        second.next = third;
        third.next = fourth;
        fourth.next = second; 
        
        String result = hasCycle(head);
        System.out.println("Result: " + result);
    }
}

File: DSA/Hima-Swarupa-Padala/day6q1.py
class ListNode:
    def __init__(self, value=0, next=None):
        self.value = value
        self.next = next

def has_cycle(head):
    if not head or not head.next:
        return "no cycle"
    
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return "cycle"
    
    return "no cycle"

a = ListNode(1)
b = ListNode(2)
c = ListNode(3)
d = ListNode(4)

a.next = b
b.next = c
c.next = d
d.next = b 

print(has_cycle(a))  
